lib/flavor-downloader.js

```diff
@@ -0,0 +1,140 @@
+var logger = require('./logger/logger');
+var config = require('./Configuration');
+var m3u8Handler = require('./promise-m3u8');
+var chunklistM3u8Generator = require('./ChunklistManifestGenerator');
+
+var FlavorDownloader = (function () {
+
+    function FlavorDownloader(m3u8Url, destPath, entryId, flavor) {
+        this.entryId = entryId;
+        this.flavor = flavor;
+        this.streamUrl = m3u8Url;
+        this.destFolderPath = destPath;
+        this.shouldRun = true;
+        this.pollingInterval = config.get("pollingInterval");
+        this.tmpM3u8Folder = path.join(destPath, 'tmpM3u8');
+        this.playlistName = path.basename(url.parse(m3u8Url).pathname);
+        this.m3u8generator = chunklistM3u8Generator(this.destFolderPath, this.playlistName, config.get("dvrWindow"));
+    }
+
+    var formatLogMsg = function (that, msg) {
+        return "entryId: " + that.entryId + " , flavor: " + that.flavor + " - " + msg;
+    };
+
+    var downloadIndefinitely = function () {
+
+        var that = this;
+        downloadPlaylist(that)
+
+            .fail(function failure(listOfErrors) {
+                logger.error(formatLogMsg(that,"Failed to download all chunks:\n"));
+                var errList = _.map(listOfErrors, function (i) {
+                    return i.reason + "\n";
+                })
+                logger.error(formatLogMsg(that,errList));
+            });
+
+        setTimeout(function () {
+            if (that.shouldRun) {
+                downloadIndefinitely.call(that);
+            }
+            else {
+                logger.info(formatLogMsg(that,'downloader has shut down'));
+            }
+        }, that.pollingInterval);
+    };
+
+    var downloadPlaylist = function (that) {
+
+        var parsedUrl = url.parse(that.streamUrl);
+        var baseUrl = parsedUrl.href.substring(0, parsedUrl.href.lastIndexOf('/'));
+        var playlistName = path.basename(parsedUrl.pathname);
+        var playlistDestPath = path.join(that.tmpM3u8Folder, playlistName);
+
+        var downloadedM3u8 = undefined;
+
+        return httpUtils.downloadFile(that.streamUrl, playlistDestPath)
+
+            .then(function finishedDownload() {
+                return m3u8Handler.parseM3U8(playlistDestPath);
+            })
+
+            .then(function listFilesInDest(m3u) {
+                logger.debug(formatLogMsg(that,"Received M3U8: " + m3u.toString()));
+                downloadedM3u8 = m3u;
+                return qfs.list(that.destFolderPath);
+            })
+
+            .then(function downloadMissingChunks(files) {
+                //get all files that are in playlist but not in dir
+                var chunksToDownload =  _.chain(downloadedM3u8.items.PlaylistItem)
+
+                    .filter(function(item) {
+                        return !_.contains(files, item.get('uri'));
+                    }).value();
+
+                //generate updated m3u8
+                that.m3u8generator.update(chunksToDownload);
```

Comment:
Each iteration is comprised of 2 steps (which can be executed in parallel once the missing chunk list is calculated) - 
1. download missing chunks
2. update missing chunks in m3u8
if any of these fails your overall function (downloadPlaylist) should succeed or fail accordingly